prov/gni/src/gnix_msg.c

```diff
@@ -54,6 +54,110 @@
 /*******************************************************************************
  * helper functions
  ******************************************************************************/
+static void __gnix_msg_unpack_data_into_iov(const struct iovec *dest,
+					    size_t dest_cnt,
+					    void *src, size_t src_len,
+					    char is_multi_recv)
+{
+	/*
+	 * TODO: Does this fn need be thread safe? Since fi_recv is
+	 * non-blocking and we have static variables below for multi_recv.
+	 * We can only release the lock once the destination buffer is
+	 * exhausted.
+	 * TODO: Is starvation a problem here?
+	 */
+	static int i;
+	static size_t cum_len;
+	static size_t offset;
+	static fastlock_t multi_recv_lock = {0};
+	size_t bwrite;
+
+	if (is_multi_recv) {
```

Comment:
> One thread might initialize, then gets the lock, then another thread initializes.

Good point!  I will remove the FI_MULTI_RECV changes from this branch but save them elsewhere in case that functionality is required in the future.